### PR TITLE
Add frameworkdirs support to gmake and gmake2 with gcc/clang toolsets…

### PR DESCRIPTION
**Pull request by noresources**
… on macOS systems

* Add new optional parameter to toolset.getincludedirs(dirs, sysdirs, frameworkdirs)
* Translate frameworkdirs to -F<path> build & linker flags
* Add tests

**Anything else we should know?**
It should be easy to update CodeLite too but I can't really test it. 

**Did you check all the boxes?**

- [x] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [x] Add unit tests showing fix or feature works; all tests pass
- [ ] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] Minimize the number of commits
- [ ] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

### CODE REVIEW

File: src/tools/gcc.lua [hunk 1]
```diff
@@ -284,12 +284,20 @@
...
+	function gcc.getincludedirs(cfg, dirs, sysdirs, frameworkdirs)
 		local result = {}
-		for _, dir in ipairs(dirs) do
+		for _, dir in ipairs(dirs or {}) do
 			dir = project.getrelative(cfg.project, dir)
 			table.insert(result, '-I' .. p.quoted(dir))
 		end
+
+		if (table.contains(os.getSystemTags(cfg.system), "darwin")
+			and cfg.frameworkdirs ~= nil and #cfg.frameworkdirs) then
```
Comment (samsinsane) [inline src/tools/gcc.lua:295]: This should be on one line. This isn't using the right variable, shouldn't be using `cfg.` for `frameworkdirs`. Also, this is not consistent with the other `dirs` blocks by checking the value of `cfg.frameworkdirs`/`frameworkdirs`.

File: src/tools/gcc.lua [hunk 2]
```diff
@@ -284,12 +284,20 @@
...
+			for _, dir in ipairs(frameworkdirs or {}) do
+				table.insert(result, '-F' .. p.quoted(dir))
```
Comment (samsinsane) [inline src/tools/gcc.lua:297]: This isn't consistent with the other blocks, the others use `project.getrelative()` on `dir`, why doesn't this one?

File: src/tools/gcc.lua [hunk 3]
```diff
@@ -284,12 +284,20 @@
 -- Decorate include file search paths for the GCC command line.
 --
 
-	function gcc.getincludedirs(cfg, dirs, sysdirs)
+	function gcc.getincludedirs(cfg, dirs, sysdirs, frameworkdirs)
 		local result = {}
-		for _, dir in ipairs(dirs) do
+		for _, dir in ipairs(dirs or {}) do
```
Comment (samsinsane) [inline src/tools/gcc.lua:289]: Is this change necessary? Is the `or {}` required on any of the blocks in this function?

File: src/tools/gcc.lua [hunk 4]
```diff
@@ -481,6 +489,13 @@
 			table.insert(flags, '-L' .. p.quoted(dir))
 		end
 
+		if (table.contains(os.getSystemTags(cfg.system), "darwin")
+			and cfg.frameworkdirs ~= nil and #cfg.frameworkdirs) then
```
Comment (samsinsane) [inline src/tools/gcc.lua:493]: This should be on one line. Also, this is not consistent with the other `dirs` blocks by checking the value of `cfg.frameworkdirs`.

**Review by samsinsane — CHANGES_REQUESTED** [review]

File: src/tools/gcc.lua (hunk 1 above)
Comment (noresources) [reply src/tools/gcc.lua:295]: Ok for the "one-line".

The use of cfg. is a copy-paste-is-evil mistake (from `gcc.getLibraryDirectories(cfg)`), that I previously fixed and re-introduced while rebasing for this PR...

The test on frameworkdirs value is made to minimize changes in code. Otherwise it will
require to add `, {}` on all `toolset.getincludedirs()` calls. I suggest to keep it for now.

File: src/tools/gcc.lua (hunk 2 above)
Comment (noresources) [reply src/tools/gcc.lua:297]: During tests I always used absolute path so i didn't catch my attention. I will made more tests to ensure relative paths are accepted by the -F option

File: src/tools/gcc.lua (hunk 3 above)
Comment (noresources) [reply src/tools/gcc.lua:289]: No, ti was (probably) made due to an invalid call of this function in a test. Test will be fixed

File: src/tools/gcc.lua (hunk 4 above)
Comment (noresources) [reply src/tools/gcc.lua:493]: Ok for the "one line". Here, `cfg.frameworkdirs` is always valid so I will remove the unnecessary tests

File: src/tools/gcc.lua (hunk 1 above)
Comment (samsinsane) [reply src/tools/gcc.lua]: > The test on frameworkdirs value is made to minimize changes in code. Otherwise it will
require to add `, {}` on all `toolset.getincludedirs()` calls. I suggest to keep it for now.

Sorry to be pedantic, but isn't that why you have the `or {}` inside? Which is how the `sysdirs` loop works too. Checking if `frameworkdirs` is not `nil` isn't necessary because you're already doing `ipairs(frameworkdirs or {})`?

**Review by samsinsane — APPROVED** [review]

File: modules/gmake/gmake_cpp.lua [hunk 5]
```diff
@@ -521,7 +521,7 @@ end
 
 
 	function make.includes(cfg, toolset)
-		local includes = toolset.getincludedirs(cfg, cfg.includedirs, cfg.sysincludedirs)
+		local includes = toolset.getincludedirs(cfg, cfg.includedirs, cfg.sysincludedirs, cfg.frameworkdirs)
```
Comment (Jarod42) [inline modules/gmake/gmake_cpp.lua:524]: Should probably be done for all exporters.

File: modules/gmake2/gmake2_cpp.lua [hunk 6]
```diff
@@ -531,8 +531,8 @@
 			end
 		end
 
-		if fcfg.includedirs or fcfg.sysincludedirs then
-			local includes = toolset.getincludedirs(cfg, fcfg.includedirs, fcfg.sysincludedirs)
+		if fcfg.includedirs or fcfg.sysincludedirs or fcfg.frameworkdirs then
+			local includes = toolset.getincludedirs(cfg, fcfg.includedirs, fcfg.sysincludedirs, cfg.frameworkdirs)
```
Comment (Jarod42) [inline modules/gmake2/gmake2_cpp.lua:535]:
```suggestion
			local includes = toolset.getincludedirs(cfg, fcfg.includedirs, fcfg.sysincludedirs, fcfg.frameworkdirs)
```

File: src/tools/gcc.lua [hunk 7]
```diff
@@ -284,12 +284,20 @@
...
-	function gcc.getincludedirs(cfg, dirs, sysdirs)
+	function gcc.getincludedirs(cfg, dirs, sysdirs, frameworkdirs)
 		local result = {}
 		for _, dir in ipairs(dirs) do
 			dir = project.getrelative(cfg.project, dir)
 			table.insert(result, '-I' .. p.quoted(dir))
 		end
+
+		if (table.contains(os.getSystemTags(cfg.system), "darwin") and frameworkdirs ~= nil) then
+			for _, dir in ipairs(frameworkdirs or {}) do
```
Comment (Jarod42) [inline src/tools/gcc.lua:295]: Below doesn't have `or {}`.

**Review by Jarod42 — CHANGES_REQUESTED** [review]

File: modules/gmake/gmake_cpp.lua (hunk 5 above)
Comment (noresources) [reply modules/gmake/gmake_cpp.lua:524]: This "miss" was intentional. I don't think resflags needs framework dirs.

File: src/tools/gcc.lua (hunk 7 above)
Comment (noresources) [reply src/tools/gcc.lua:295]: `and frameworkdirs ~= nil` is probably the one to remove here

File: modules/gmake/gmake_cpp.lua (hunk 5 above)
Comment (Jarod42) [reply modules/gmake/gmake_cpp.lua:524]: For resources, it might be OK, but I would pass `nil` explicitly.

For gmake2, only for "per-file" configuration has been done.
Missing for Codelite generator...

Comment (noresources) [reply modules/gmake/gmake_cpp.lua:524]: I am not against an explicit nil, however there is currently no `toolset.getincludedirs(cfg, inc, nil)` when sysdir is not mandatory.
About gmake2, Ok, I didn't understand your first message quite well. This will be fixed.
About CodeLite, I said in the PR message I am not using CodeLite, have no knowledge about it and I can't test it.

Comment (Jarod42) [reply modules/gmake/gmake_cpp.lua:524]: `sysdir` is missing only for resources flags (I wonder it is not a bug BTW).

You should add `frameworkdirs` for all non-resources flags call (for all generators).

It is the toolset which "manages" it. (as msvc which does nothing).
if generator doesn't use directly compiler flags, it iterates directly on `cfg.includedirs` or similar, and then indeed, it would require knowledge of the generator.

For test, I started https://github.com/Jarod42/premake-sample-projects to test with projects with multiple generators.
Nothing platform specific yet, as `frameworks` 

If you can provide sample test project (usable in github action) (a variant of project-01 which test includedirs), I can add it.

**Comment by noresources** [comment]
Hello,
I finally had some spare time and motivation to run a macOS VM, install CodeLite and test frameworkdirs  support.
I also add the runpathdirs support to it (one line change).
Finally I rebased/cleanup commits on top of master.
Hope this will allow to close this PR.

File: modules/gmake/gmake_cpp.lua (hunk 5 above)
Comment (starkos) [reply modules/gmake/gmake_cpp.lua:524]: @Jarod42 It is okay if contributors only want to target a single toolset. Not everyone has access to every toolset; allowing people to contribute where they can is more important. And please do not confuse things by asking people to provide you with tests for a separate repository; unit tests are all that we require at this time. Thanks!

**Review by starkos — APPROVED** [review]
Approving, but will let @samsinsane make the call on the merge since they've been more involved in the review.

**Comment by starkos** [comment]
Oh shoot, I see @samsinsane already approved…merging…

File: modules/gmake/gmake_cpp.lua (hunk 5 above)
Comment (Jarod42) [reply modules/gmake/gmake_cpp.lua:524]: @starkos: What bothers me is that, then projects are generator-specific without any errors/warnings for users when used with incompatible generators.

Sorry, I didn't want to confuse people with extra tests in another repository.

Notice that you have tests just for gcc, but not for gmake/gmake2.

Comment (starkos) [reply modules/gmake/gmake_cpp.lua:524]: > What bothers me is that, then projects are generator-specific without any errors/warnings for users when used with incompatible generators.

The alternative is that people don't contribute. We've been down that path before. I do agree that we need to do a better job of surfacing compatibility information, and the documentation seems an appropriate place to do that.

@Jarod42 If you have general comments on broad topics ("features should be implemented for all generators") please use the provided Discussions rather than cluttering up the PR review notes. I have no longer have any idea what change was requested by this thread, and if it has been done or not.

File: src/tools/gcc.lua (hunk 7 above)
Comment (starkos) [reply src/tools/gcc.lua:295]: I believe this one has been resolved?
